tensorflow_zero_out/cc/kernels/lsh_matmul.cc

```diff
@@ -0,0 +1,221 @@
+#define BISINIT 0
+#define BINDEX 1
+#define BCOUNT 2
+//TODO: fill the values below
+#define BUCKETSIZE 
+#define _L
+#define _numhashes 
+#define _noOfNodes 
+
+  int Bucket_add(int id, int l, int idx, int L, int RangePow, int BUCKETSIZE,
+                 double *lsh_table) {
+
+  int getHashSparse(int* indices, float *values, int length) {
+    int *hashes = new int[_numhashes];
+
+    for (int p = 0; p < _numhashes; p++) {
+        double s = 0;
+        int i = 0;
+        int j = 0;
+        while (i < length & j < _samSize) {
+          if (indices[i] == _indices[p][j]) {
+            float v = values[i];
+            if (_randBits[p][j] >= 0) {
+              s += v;
+            } else {
+              s -= v;
+            }
+            i++;
+            j++;
+          }
+          else if (indices[i] < _indices[p][j]){
+            i++;
+          }
+          else{
+            j++;
+          }
+        }
+        hashes[p] = (s >= 0 ? 0 : 1);
+    }
+    return rawResults;
+  }
+
+    return hashes;
+  }
+
+  int* hashesToIndex(int * hashes) {
+    int * indices = new int[_L];
+    for (int i = 0; i < _L; i++) {
+      unsigned int index = 0;
+      for (int j = 0; j < _K; j++) {
+        if (HashFunction==4) {
+	  unsigned int h = hashes[_K*i + j];
+	  index += h<<(_K-1-j);
+	} 
+      indices[i] = index;
+      }
+    } 
+    return indices;
+  }
+
+  int** retrieveRaw(int *indices) {
+    int ** rawResults = new int*[_L];
+    int count = 0;
+    for (int i = 0; i < _L; i++)
+      rawResults[i] = _bucket[i][indices[i]].getAll();
+    return rawResults;
+  }
+
+  int getCandidates(int **actives) {
+    for(int i = 0; i < _L; i++) {
+      if(actives[i] == NULL) {
+        continue;
+      } 
+      else {
+        for(int j = 0; j < BUCKETSIZE; j++) {
+          int tempID = actives[i][j] - 1;
+          if(tempID >= 0) 
+            counts[tempID] += 1;
+          else
+            break;
+        }
+      }
+    }
+
+    if(counts.size()<1500){
+      srand(time(NULL));
+      int start = rand() % _noOfNodes;
+      for(int i = start; i < _noOfNodes; i++) {
+        if(counts.size() >= 1000) 
+          break;
+        if(counts.count(_randNode[i]) == 0)
+          counts[_randNode[i]] = 0;
+      }
+
+
+      if(counts.size() < 1000) {
+        for(int i = 0; i < _noOfNodes; i++) {
+          if(counts.size() >= 1000)
+            break;
+          if(counts.count(_randNode[i]) == 0)
+            counts[_randNode[i]] = 0;
+        }
+      }
+    }
+
+    len = counts.size();
+    return len;
+  }
+};
+
+void Compute(OpKernelContext* ctx) override {
+    // Note:
+    // need values(x), weights, indices, inputID,  activenodesPerLayer, 
+    // activeValues & length as input parameters or attributes
+    const Tensor& a = ctx->input(0);
+    Tensor an = ctx->input(1);
+    Tensor av = ctx->input(2);
+    Tensor l = ctx->input(3);
+    Tensor i = ctx->input(4);
+    Tensor w = ctx->input(5);
+    Tensor in = ctx->input(6);
+
```

Comment:
It looks like these tensor need to be either `const Tensor&` or `Tensor&` depending on if we need to modify them.